enhancements/authentication/configuring-webhook-token-authenticators.md

```diff
@@ -0,0 +1,176 @@
+  in `webhookTokenAuthenticators` field
+- it pushes any webhook authenticator that passes the validation to `openshift-kube-apiserver/webhooktokenauthenticators`
+  secret, where `key=authenticator_name` where `authenticator_name` is either the name of the referenced secret or
+  the CR
+- if no webhook authenticators pass validation, the operator keeps the previous state
+- redeploys the kube-apiserver static pods while specifying the new
+  mounted files in the `webhookTokenAuthenticators` [config field](https://github.com/openshift/api/blob/992bc3a41fe6c4e5bcf27dc8080f9e5c114c33fc/kubecontrolplane/v1/types.go#L146).
+
+#### Backend - Integrated OAuth Server Configuration
+
+It's cluster-authentication-operator's responsibility to:
+- observe `authentication.config/cluster` field
+- if set to "IntegratedOAuth", it creates secret `openshift-config/integrated-oauth` and adds it to
+  the `webhookTokenAuthenticators` field of the `authentication.config/cluster` resource
+- the above secret contains details that point kube-apiserver to the oauth-apiserver for token reviews
+
+To perform token reviews for tokens issued by the integrated oauth-server, the new oauth-apiserver needs to
+be able to handle token review requests, such capability is currently missing and needs to be implemented.
+
+When responding after a successful authentication, the oauth-apiserver needs to return all of the users
+groups, on top of which it needs to add the `system:authenticated:oauth` group to keep the behavior of
+previous verions. This group adds self-provisioner cluster role to the authenticated user, the role
+allows users privileges necessary to work with projects.
+
+Similarly, if any integrator would like their users to have project privileges, they need to add this
+specific group in their token review responses.
+
+### Risks and Mitigations
+
+There are a couple of issues that root in the very fact of making it possible of
+replacing the oauth-server with a custom token authenticator:
+
+- `oauth-proxy` relies on the _/.well-known/oauth-authorization-server_ endpoint
+  and if the integrated oauth-server is turned off and the `oauthMetadata` is not set
+  properly, things like grafana and prometheus won't be accessible
+- function of selfubjectaccessreviews and the "users/~" echo endpoint may not work
+  properly
+- ServiceAccounts may not be usable as OAuth clients anymore
+
+It is a sole responsibility of the replacing part to make sure all of these work.
+
+## Design Details
+
+### Test Plan
+
+Since we will be making the cluster-authentication-operator use the `webhookTokenAuthenticatros`
+(in combination with `oauthMetadata`, which already works) to configure authentication using
+the integrated OAuth server, we will just re-use all the tests that we have already.
+
+### Upgrade / Downgrade Strategy
+
+A short period of OpenShift token authentication is perhaps unavoidable as it is
```

Comment:
I don't see why you would ever become unavailable.  You own both operators and can have a release of overlap if necessary.  Give it some more thought.  Hint: you can keep the old oauth token authenticator until a webhook token authenticator setting is present and you control when that value is set.  In 4.5, you no longer have to keep the old flow.

Reply:
If it's ok to keep the KAS patching code until 4.5, this should indeed be as easy as `if len webhooks == 0 still patch it`, in that case we wouldn't probably even have to do any escapades in 4.3 CAO

Reply:
Modified this section accordingly